### ISSUE TITLE
Feature request ripple stitch for fill

### ISSUE THREAD

**Issue by Tina-Chris** [inkstitch/inkstitch#1196]
Hey; I checked before askting this:
Inkscape does not allow us to spiral/ripple fill our shapes so far. In the Inkscape Forum I asked if I just missed a function (https://inkscape.org/forums/beyond/spiral-filling-any-shape/), but they seem to agree, that it is currently necessary to do it manually.
Ken Wilson suggested the use of the following script; https://www.geeksforgeeks.org/draw-shape-inside-shape-in-python-using-turtle/
I would imagine a tab as the fill tab on top where we have a ripple fill tab to select, which runs this turtle fuction. Would this be possible? Maybe similar to the gradient fill? Or would it be more suitable to have it integrated in the Inkscape itself?

As a way to explain, I mean to fill a heart / leaf / star with a spirally loose shape: See here on the 1st heart as an example:
![image](https://user-images.githubusercontent.com/64272979/117506414-d7257700-af85-11eb-8a0b-0b92c261d608.png)
Image taken from hatch https://www.embroideryhelp.net/hatch/digitizer/en/user-guide/stitch-types/curved-fills 
Thanks
Tina
This might go along the lines of https://github.com/inkstitch/inkstitch/issues/692 and https://github.com/inkstitch/inkstitch/issues/33 but with having a python ready made script to use as basis for the fill?

**Comment by lexelby** [comment]
I'd love to have this feature!  That's an interesting link, but it seems as if that code is only designed to handle triangles, squares, and hexagons.  We'd have to come up with generalized code that could handle any shape.

That won't be completely straightforward.  Imagine a bowtie shape: ⧓


At the start, it would be fine, but very soon, the center part would meet.  Then what do we do?  The spiral would go to the center of one side and then get stuck, and half of the shape would be unfilled.

**Comment by B-ellanna** [comment]
For a bow tie we have the same problem with the current fill stitch, as it can not in all angles always pass through the narrow section. So I would rather have a script which can handle basic cases, and do the rest manually, than having to do all manually :).

**Comment by lexelby** [comment]
Fair enough. :)

**Comment by AndyLV** [comment]
Hello together,
some months ago I started to develop another fill method using offset curves. These curves are afterwards connected and re-sampled to get an interlaced point structure (as best as possible so far):
![Interlaced](https://user-images.githubusercontent.com/45293380/125335293-b0503c00-e34c-11eb-8760-7bbbc5924229.png)
The input is a boundary shape which also can have holes. There are still some work packages open like dealing with holes remaining after the recursive offset - e.g.
![known_gaps](https://user-images.githubusercontent.com/45293380/125335624-0cb35b80-e34d-11eb-987d-f9183b08fd49.png)
but this should cause gaps only at certain positions.
The code is developed in Python3 making heavily use of Shapely to be as compatible as possible for an integration in Inkstitch later on if favored.
Nevertheless, after some months of development I wanted to give it a try on an embroidery machine. I used a rabbit whose stitched result with the existing method looks like this:
![Original](https://user-images.githubusercontent.com/45293380/125335961-703d8900-e34d-11eb-88d0-c9b48d27cf52.jpg)
However, the first stitched version looked rather bad:
![Draft1](https://user-images.githubusercontent.com/45293380/125336052-93683880-e34d-11eb-9671-a39b811e4b1e.jpg)
(The belonging file generated by my Python script: https://homepages.uni-paderborn.de/anlu/Draft1.svg - I opened it with Inkscape, clicked on Ink/Stitch->Parameter to create a manual stitch along a path. In addition this method is very helpful to get an animation of the stitch order. The resulting embroidery file is: https://homepages.uni-paderborn.de/anlu/Draft1.vp3).
Since the result looked so bad I assumed that a potential problem was that adjacent lines were not always stitched directly consecutive (as you can see when animating Draft1.svg). Hence I rewrote my script forcing it to strictly stitch from inner to outer. The result looks like this:
![Draft2](https://user-images.githubusercontent.com/45293380/125337026-c3640b80-e34e-11eb-879a-61ec41c23ba1.jpg)
(belonging files: https://homepages.uni-paderborn.de/anlu/Draft2.svg and https://homepages.uni-paderborn.de/anlu/Draft2.vp3).
As you can see the result marginally improved. However, to be honest - I'm still disappointed about the huge number of gaps.
Since I'm not so experienced in stitching and I haven't used a professional embroidery software so far - may I ask you for some hints how the points need to be arranged to avoid these gaps? (Of course an underlay would minimize it but since I didn't use an underlay when stitching with the existing method, I want to achieve a comparable quality with this offsetting method without underlay.)
When importing the generated vp3 files within Inkscape I noticed a difference between the stitch plan and what was saved to the vp3 file:
![Comparison_vp3](https://user-images.githubusercontent.com/45293380/125337813-b85dab00-e34f-11eb-9e66-a2ff95151377.png)
Maybe this explains some of the seen gaps - but what is the reason for this deviation?

Thank you very much for your support and kind regards,

Andreas

**Comment by Tina-Chris** [comment]
Hi Andreas
To check your last question.
You might try to save as pes file, and then export/convert with another
embroidery software to create a vp3.
Let's see if this gives the same result?
Thanks for taking the time to make this code!
Tina

AndyLV ***@***.***> schrieb am Mo., 12. Juli 2021, 20:35:

**Comment by lexelby** [comment]
COOL!!!

Great work on this.  I'm super impressed and I can't wait to get this into Ink/Stitch.

I have a couple of comments/ideas.  I'll start with the last thing first: I think the weirdness you're seeing when loading the VP3 is because VP3, like most embroidery file formats, has a resolution is 0.1mm.  That's fine, because the reality of fabric and thread means that any finer accuracy would be irrelevant anyway.

That may not be the whole story though.  Try the same thing with PES and DST.  It's very possible we have some kind of error in our VP3 export code.

Next up: if your goal is to avoid those gaps, underlay is really important.  I think it's going to make a world of difference. Stitch order will too, and increasing the density.  I'd definitely start with the underlay though.

But... I think eliminating gaps is not necessary to produce a really neat effect.  If we do your algorithm with wider spacing, it will look quite purposeful, and it will be a striking visual effect. So I would love to incorporate this code even as it is now.

**Comment by AndyLV** [comment]
Hello together,

I tested the export of different formats as well as the conversion to vp3:
![Comparison](https://user-images.githubusercontent.com/45293380/125592022-ad511462-1aa1-48e9-97ac-a1131fceffd0.png)
For the conversion from dst to vp3 I used this tool: https://chrome.google.com/webstore/detail/embroidery-viewer-and-con/dhecljdbpaninichncjocoljjihkfacj?hl=de

In my opinion the converted results looks better than the natively generated vp3 file by Ink/Stitch.

Regarding the code - I pushed the current state here:
https://github.com/AndyLV/tangential_fill_stitch

One important remark since you mentioned that you want to use it with wider spacing: The original intention was to use this script to generate a covering pattern fill hence the spacing was seen as a measure for very close points. Therefore, points which are closer than factor*spacing are usually removed. If you now increase the spacing, your geometry might be more simplified than you intended. If this is the case you need to edit the file https://github.com/AndyLV/tangential_fill_stitch/blob/develop/constants.py and decrease the constants which are documented as factors for the used spacing.

If you have any problems or questions do not hesitate to ask.

Kind regards,

Andreas

**Comment by AndyLV** [comment]
Regarding the vp3 format I had a look on the source code:

In https://github.com/inkstitch/pyembroidery/blob/235e5d044bc0913b2a01758935151e10d3e1db49/pyembroidery/Vp3Writer.py
line 248 & 249 you cast the stitch delta x,y to _int_ (effectively _floor_) while in https://github.com/inkstitch/pyembroidery/blob/235e5d044bc0913b2a01758935151e10d3e1db49/pyembroidery/DstWriter.py you call _round()_ before (Line 165 & 166).
I don't know whether this explains alle differences in my last post but I think introducing a round() in the Vp3Writer would be beneficial. If line 248 & 249 will be changed I would also consider lines 119,120,126-132,177,178,185 and 186 whether introducing a round() would be useful.

**Comment by lexelby** [comment]
Yup, you're right!  Definitely a problem in our VP3 code.  Our pyembroidery was written by a former Ink/Stitch contributor who took issue with our code of conduct, and unfortunately we haven't dedicated much time to improving it since their departure.  They may well have fixed this issue in their version, embroidepy/pyembroidery.  If I have a chance, I'll take a look, and of course we would be very interested in pull requests.

**Comment by lexelby** [comment]
Wow, that's a lot of code. :)  It's going to take me awhile to understand it.

If you're interested in integrating it into Ink/Stitch, how would you like to proceed?  One option would be for us to give you write access to this repository.  That would give you the ability to push a branch and create a pull request. When you do that, our build system will automatically post a release on our releases page to make testing easy.

That would probably be the best method if you're interested.  Otherwise, if you'd rather we do the integration work, we can, but it may take awhile as we're somewhat low on time and working on other tasks at the moment.

One thing we'll need to work out is licensing.  I see that you're using a python file taken from a gist.  Since that code doesn't have a license associated with it, unfortunately we would not be able to integrate it into Ink/Stitch without explicit permission from its author.  Fortunately, I don't think that would be necessary, since we already have code to take shapes from an SVG and convert curves into LineStrings.

**Comment by AndyLV** [comment]
Regarding the vp3:
I downloaded the newest version from https://github.com/EmbroidePy/pyembroidery/tree/master/pyembroidery to test the vp3 export - it gave me this result:
![newPyembroidery](https://user-images.githubusercontent.com/45293380/126547198-eace289b-5b35-42a0-bb7c-695e29781253.png)
(From my perspective the problem was not solved).
I introduced the round() in line 255 & 256 (different lines than mentioned in my last post since I used the newest version of pyembroidery (not the branch used by inkstitch) now) - in my opinion this gives much better results:
![newPyEmbroideryRound](https://user-images.githubusercontent.com/45293380/126547540-44ea03a5-eeda-42c9-a31d-547e40e50358.png)
If I'm correct I currently can not create a pull request for the fix since I have no write access?

Regarding the integration in Ink/Stitch (to avoid integration effort on either side for a feature you expected to be different): Since my original intention (pattern fill) was different than your use case (wide spacing), I want to reassure that the code produces what you really want. E.g. for larger spacings the rabbit looks like (the constants.py file was slightly modified to reduce the simplification):
![LargerDistance](https://user-images.githubusercontent.com/45293380/126548048-9f4c5dad-cd83-4144-a8e8-6226afe20d6a.png)
I want to stress that the code does not produce a connected spiral like in Tina's first image (since a simple spiral can not cover more complex shapes where the offsetting method produces more than one "leaf" (closed shape without any more inner shapes)). Hence the code introduces several bridges (marked in black) in order to enter and leave a sub geometry at the same place.  Is this still want you want?

Regarding the license - I fully agree. I took this additional file only for my tests. As soon as the code would be integrated into Ink/Stitch this file would become obsolete.

One last remark regarding your manual setup instruction page (https://inkstitch.org/developers/inkstitch/manual-setup/):
Maybe someone else can test it and confirm or disagree but I encountered several problems on Ubuntu and Linux Mint:

1) In Step "2. Install Pyembroidery" you did not "cd" in the directory of inkstitch before the clone, so pyembroidery will become a folder on the same level as inkstitch. This causes a problem in step "3. Python DependenciesPermalink" where you call "pip install -r requirements.txt" inside the folder inkstitch. The file requirements contains a line "./pyembroidery" which is not a valid path in this case. In order to fix it you have two possibilities: 1) clone pyembroidery into the inkstitch folder or 2) change the line in the requirements file to "../pyembroidery".

2) In Step 4 I had to use "yarn install --ignore-engines" instead of "yarn install" - depends probably on the machine but might be a good hint.

3) Step 5 worked only for me if I changed the symbolic link "python" from python 2 to python 3. Maybe one can introduce a shebang or call the script explicitly with phython 3 to avoid this?

Kind regards,

Andreas

**Comment by lexelby** [comment]
That does look much better!  Thanks for doing the legwork to verify this fix.  I've granted you write access to this repo and pyembroidery so that you can push and create a PR.

I gave the wrong impression above, that I would only want your code for wider spacing.  That's far from the case!  I think folks will be very interested in using the feature as you intend it too.  

But don't underestimate the creativity of our users.  Even though, as you showed, it doesn't quite do what I was thinking at wider spacing, I bet folks will find a use for that too.  IMO we should give them the ability to configure the constants and let them experiment.

I also think it's worth integrating now, even before you've gotten the fill last the way how you want it.  It's already cool!

Bottom line: yes please, let's integrate.  I know you're not the only one interested in this kind of fill stitch.

Thanks for your detailed feedback on the manual install docs.  They definitely need some work.  If you're interested in working on them, the website docs are in the `gh-pages` branch in this repo. If not, we can create a new issue with your notes so that someone can come along and take on that work.

**Comment by AndyLV** [comment]
Thanks for giving me the write access - I added a pull request to pyembroidery which was also accepted, now.

In addition, I made a first try to implement the tangential fill feature into Ink/Stitch - the branch can be found here:
https://github.com/inkstitch/inkstitch/tree/feature_tangential_fill

Since this adds a third fill method to the current implementation (next to auto_fill and "legacy"/manual fill) the toggle mechanism between the tabs of auto_fill and manual fill was difficult to retain (since a toggle state allows only two methods). That is why I removed the toggling between the tabs and added a choice widget. Based on the selection of the choice widget further widgets are shown or hidden. To get a better impression I made a screen recording (the settings are in German but I think you can imagine what is going on):

https://user-images.githubusercontent.com/45293380/127882704-57a2be5e-b6c0-4f3d-aab7-eadefebd21ba.mp4

Since it is an old laptop and the tangential fill method is not fully optimized right now it takes some seconds to calculate the solution.

Regarding the manual install docs - I would prefer that a new issue will be created for this so that the best solution for the setup can be found.

**Comment by Tina-Chris** [comment]
Hi Andy!
Thanks for enriching the team with your time and knowledge :)!
It looks like a cool thing to try.
Tina

**Comment by kaalleen** [comment]
Wow, nice. It's looking good.

**Comment by Augusa** [comment]
Really looking forward to being able to use all of this. Thank you

**Comment by lexelby** [comment]
Awesome!! I'm diving into the code now.  I think it's going to take me some time to understand it :)

I like what you did with the fill types, both in the GUI and the text.  What you've done matches well with what I have been working on in #798.  #798 is going to be a big change that gets rid of the wxPython Params GUI entirely and re-implements it in Electron.

As part of working on #798, I realized the same thing as you: having Fill and AutoFill as separate classes was confusing and limiting.  I also combined the two in my branch, and I went even further and combined Stroke and SatinColumn too.

I'm hoping to have some time to work on #798 this week.  If I can get it finished, I think it's going to make it much easier to integrate your stuff.

**Comment by AndyLV** [comment]
I'm currently working on rewriting the LineStringSampling.py file (+refactoring of the whole code to make it easier to understand)  since I'm sometimes not satisfied how the points are placed - so maybe you wait with your code dive ;-)
Regarding the new GUI using Electron - is it possible to have something like a progress bar? Since the point placing routines are quite computational expensive a progress bar might prevent the user of thinking the application crashed.

**Comment by lexelby** [comment]
I'll be honest with you, I already tried twice to understand the code, starting from different places, and I just couldn't wrap my head around it. ;)  I'm glad to hear you're reworking it.  Thank you!  

Could I ask one minor favor?  We prefer to have file names and variable names in snake_case for consistency.  CamelCase is preferable for classes.  Thanks!

Yes, a progress bar is definitely in my mind. Easy to do in Vue + Vuetify, and we already have a spinner for the simulator.

You can see what I've got so far in #798.  Still quite a bit to do, of course :)

**Comment by AndyLV** [comment]
Just to give a short update:
As already mentioned I was not satisfied how the points were placed - e.g. the previous  method needed 3496 stitches for the ellipse:
![previousVersion](https://user-images.githubusercontent.com/45293380/134050251-a9533355-a438-4aca-8556-e815cd151368.png)

For comparison - the "normal" autofill method needed only 1209 stitches:
![AutoFill](https://user-images.githubusercontent.com/45293380/134050367-c5307014-7619-474d-8880-71c5c59c26e8.png)

The new implementation needs at least only 1970 stitches:
![newVersion](https://user-images.githubusercontent.com/45293380/134050530-698f6472-e3a3-4827-b740-56144cf9f4ee.png)

The diagonal point accumulation in the top left is due to the avoidance of overlapping segments when stitching from inner to outer:
![Entering](https://user-images.githubusercontent.com/45293380/134050688-55779fac-b5c6-4b2e-8a49-4348cf6816a6.png)
Maybe I should remove this to avoid this point accumulation.

@lexelby: I'm still working on making the code easier to understand - do you have a date in mind when your replacement of wxPython will be ready to be merged?

**Comment by kaalleen** [comment]
> I'm still working on making the code easier to understand - do you have a date in mind when your replacement of wxPython will be ready to be merged?

We do not have a date yet. There is still a lot of work to do. So I am guessing that it will take quite some time. Sorry for the delay.

**Comment by AndyLV** [comment]
As a further step towards to original question of this issue.
https://github.com/inkstitch/inkstitch/tree/feature_guided_fill


https://user-images.githubusercontent.com/45293380/143292566-4f58d085-07b9-45e0-ad18-85c2ae578f0e.mp4



Note that the outer shape needs to able to be filled by a single spiral (e.g. recursively offsetted curves must collapse to one single center)

**Comment by Tina-Chris** [comment]
@AndyLV Das sieht ja echt schon ganz toll aus! Bin gespannt :)

**Comment by Augusa** [comment]
What a wonderfull result. Hope we get it soon in a next release of Inkstitch

**Comment by kaalleen** [comment]
@AndyLV could you make your branch a pull request? I think it will be easier to find and to remember.
When we finally released the next version, I think your branch should be in focus next.
